### PR TITLE
added persistent option to buffers and namedbuffers

### LINKED ISSUE #85235: Add `persistent` option to `nn.Module.buffers`.
### 🚀 The feature, motivation and pitch

For convenience, it would be nice if the [`nn.Module.buffers`](https://pytorch.org/docs/stable/generated/torch.nn.Module.html#torch.nn.Module.buffers) could distinguish between persistent and non-persistent buffers.

```python
def buffers(self, recurse: bool = False, persistent: Optional[bool] = None) -> Iterator[Tensor]:
```

Which yields an iterator for either only persistent buffers (`True`), non-persistent buffers (`False`) or both (`None`).

### Alternatives

_No response_

### Additional context

Currently, one either has to make use of the private attribute `nn.Module._non_persistent_buffers_set`, or filter `self.buffers` via the entries of the `state_dict`.

cc @albanD @mruberry @jbschlosser @walterddr @kshitij12345 @saketh-are

### PR DESCRIPTION
**Pull request by randolf-scholz**
Fixes #85235

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @chauhang @d4l3k @albanD

### CODE REVIEW

**Review by mikaylagawarecki — COMMENTED** [review]
Looks good, will merge if CI passes.

Question: I noticed that the initial proposal in https://github.com/pytorch/pytorch/issues/85235 proposed

> Which yields an iterator for either only persistent buffers (True), non-persistent buffers (False) or both (None). 

However here we only have persistent buffers (True) and both (False). Why did we choose to drop the option to get non-persistent buffers?

File: torch/nn/modules/module.py [hunk 1]
```diff
@@ -2649,13 +2654,16 @@ def named_parameters(
...
+        self, recurse: bool = True, *, persistent: bool = False
+    ) -> Iterator[Tensor]:
         r"""Return an iterator over module buffers.
 
         Args:
-            recurse (bool): if True, then yields buffers of this module
+            recurse (bool, optional): if True, then yields buffers of this module
                 and all submodules. Otherwise, yields only buffers that
                 are direct members of this module.
+            persistent (bool, optional): if True, then yields persistent buffers only. Defaults to False.
```
Comment (mikaylagawarecki) [inline torch/nn/modules/module.py:2666]: can we make this an enum with 3 values that returns either `PERSISTENT`, `NON_PERSISTENT` or `ALL`?

Comment (randolf-scholz) [reply torch/nn/modules/module.py:2666]: @mikaylagawarecki I pushed the 3-valued options as a separate PR: #132994

**Comment by mikaylagawarecki** [comment]
Closing in favor of the new PR https://github.com/pytorch/pytorch/pull/132994

**Pull request closed without merging**